src/GitTools.jl

```diff
@@ -164,4 +167,96 @@ function fetch(repo::LibGit2.GitRepo, remoteurl=nothing; header=nothing, kwargs.
     end
 end
 
+
+# This code gratefully adapted from https://github.com/simonbyrne/GitX.jl
+@enum GitMode mode_dir=0o040000 mode_normal=0o100644 mode_executable=0o100755 mode_symlink=0o120000 mode_submodule=0o160000
+Base.string(mode::GitMode) = string(UInt32(mode); base=8)
+Base.print(io::IO, mode::GitMode) = print(io, string(mode))
+
+function gitmode(path::AbstractString)
+    if isdir(path)
+        return mode_dir
+    elseif islink(path)
+        return mode_symlink
+    elseif Sys.isexecutable(path)
+        return mode_executable
+    else
+        return mode_normal
+    end
+end
+
+"""
+    blob_hash(path::AbstractString)
+
+Calculate the git blob hash of a given path.
+"""
+function blob_hash(path::AbstractString, HashType = SHA.SHA1_CTX)
+    ctx = HashType()
+    datalen = filesize(path)
+
+    # First, the header
+    SHA.update!(ctx, Vector{UInt8}("blob $(datalen)\0"))
+
+    # Next, read data in in chunks of 4KB
+    buff = Vector{UInt8}(undef, 4*1024)
+    open(path, "r") do io
+        while !eof(io)
+            num_read = readbytes!(io, buff)
+            update!(ctx, buff, num_read)
+        end
+    end
+
+    # Finish it off and return the digest!
+    return SHA.digest!(ctx)
+end
+
+"""
+    tree_hash(root::AbstractString)
+
+Calculate the git tree hash of a given path.
+"""
+function tree_hash(root::AbstractString, HashType = SHA.SHA1_CTX)
```

Comment:
Is there a need to think about `.gitignore` here`?

Reply:
Yes, we do if we want this to actually be a "git" tree hash.

Reply:
I think what we should do is have `tree_hash()` take in a `honor_gitignore::Bool = true` but then for all artifact-related usage, we set `honor_gitignore` to `false`.  This way, we can still use this to verify packages on-disk in the future (so as to match what `git` itself thinks the git tree hash is) without having weird git-specific behavior within artifacts.

Reply:
I'm punting on this; eventually I'd like to use `tree_hash()` to also verify julia packages on-disk, but that would require proper `.gitignore` handling, which is a little more work than I want to put into this PR.  I spent an afternoon trying to make a simple parser, but `.gitignore` files are... flexible.  ;)

Reply:
Right, I was thinking in terms of verifying that what is on-disk has not changed since you installed it.